pkg/controller/route/route_controller.go

```diff
@@ -279,8 +344,32 @@ func (rc *RouteController) isResponsibleForRoute(route *cloudprovider.Route) boo
 	for i := range lastIP {
 		lastIP[i] = cidr.IP[i] | ^cidr.Mask[i]
 	}
-	if !rc.clusterCIDR.Contains(cidr.IP) || !rc.clusterCIDR.Contains(lastIP) {
-		return false
+
+	// check across all cluster cidrs
+	for _, clusterCIDR := range rc.clusterCIDRs {
+		if clusterCIDR.Contains(cidr.IP) || clusterCIDR.Contains(lastIP) {
+			return true
+		}
+	}
+	return false
+}
+
+// checks if a node owns a route with a specific cidr
+func hasRoute(rm map[types.NodeName][]*cloudprovider.Route, nodeName types.NodeName, cidr string) bool {
+	if routes, ok := rm[nodeName]; ok {
+		for _, route := range routes {
+			if route.DestinationCIDR == cidr {
+				return true
+			}
+		}
 	}
-	return true
+	return false
+}
+
+//strips : . /
+func cidrtoRfc1035(cidr string) string {
+	cidr = strings.ReplaceAll(cidr, ":", "")
```

Comment:
This feels very ambiguous, particularly with the IPv6 0 elision (e.g. can't I have `a:b::` and `ab::`)?

For IPv6, isn't the string form also not a 1:1 mapping as well, e.g. `a:b:0::` vs `a:b::`

Reply:
it does not really matter. What matters is a unique string is generated out of the ipv6. It is really meant to generate unique `nameHints`

Reply:
Right, but I'm not sure it is unique.  What if we had `a:b::` and `ab::` ?

Reply:
I get your - valid - point about uniqueness. This is not meant to be unique across the entire system just for a single node. A node will not have `a:b::` and `ab::` as two assigned CIDRs.

Reply:
At least on GCE, it is used to construct a unique name:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/gce/gce_routes.go#L77

Reply:
Ah - scratch that!  It is combined with the node UID!  I guess this is OK then, but this still feels like a big footgun!

Reply:
(The safer thing in my mind would be to parse the cidr with net.ParseCIDR, and then build it deterministically i.e. without coalescing zeroes.)